Apply schema inclusion/exclusion filtering to sql-jdbc drivers

jeff303 opened the pull request:
Update sql-jdbc namespaces to handle schema inclusion/exclusion patterns when filtering schemas

Add new generic schema inclusion/exclusion test for sql-jdbc drivers that define the property

Update Snowflake and Redshift driver manifests to include schema filtering property

jeff303 committed:
Apply schema inclusion/exclusion filtering to sql-jdbc drivers

WORK IN PROGRESS

Update sql-jdbc namespaces to handle schema inclusion/exclusion patterns when filtering schemas

Add new generic schema inclusion/exclusion test for sql-jdbc drivers that define the property

Update Snowflake and Redshift driver manifests to include schema filtering property

TODO: figure out if non-working Snowflake "schema" property should be migrated to new pattern (i.e. #13126)

jeff303 commented:
- WORK IN PROGRESS

TODO:
- [ ] add new property to other JDBC drivers as mentioned in product doc
- [ ] figure out if non-working Snowflake "schema" property should be migrated to new pattern (i.e. #13126)

jeff303 committed:
Create db-details->schema-filter-patterns util fn to turn DB details into the inclusion/exclusion patterns

Fix Snowflake to use inclusion/exclusion patterns

jeff303 committed:
Move schema inclusion/exclusion filtering code to new namespace (since it's not strictly used by :sql-jdbc derived drivers)

Move tests accordingly

Update references to new namespace

Add schema inclusion/exclusion check to the syncable-schemas multimethod (and updating docstring)

jeff303 committed:
Fix namespace decl

jeff303 committed:
Fix Redshift failure

jeff303 committed:
Fix syncable-schemas-test

Fix indentation

jeff303 committed:
Merge branch 'master' into filter-schemas-jdbc-drivers

jeff303 committed:
Merge branch 'master' into filter-schemas-jdbc-drivers

dpsutton commented on src/metabase/driver/sync.clj:
```diff
@@ -0,0 +1,51 @@
+  (:import java.util.regex.Pattern))
+
+(defn- schema-pattern->re-pattern ^Pattern [schema-pattern]
+  (re-pattern (-> (str/replace schema-pattern #"(^|[^\\\\])\*" "$1.*")
```
this could use a comment to explain the regex.

jeff303 replied:
Added in next push.

dpsutton commented on src/metabase/driver/sync.clj:
```diff
@@ -0,0 +1,51 @@
+    (filter-fn schema-name)))
+
+(defn db-details->schema-filter-patterns
+  "Given a `prop-nm` (which is expected to be a connection property of type `:schema-filters`), and a `database`
```
is it ever not "schema-filters" right now? Is this open lookup in use now or just possible future use.

jeff303 replied:
No, it's always supposed to be of type `schema-filters`, which is the custom type I added in the previous BigQuery patch to perform this logic.  I just don't think it's needed to validate it at this point (since this function will only ever be called for that conn-prop type).  Does that make sense, or did I miss your point?

dpsutton commented on src/metabase/driver/sql_jdbc/sync/describe_database.clj:
```diff
@@ -89,32 +92,45 @@
+                     (and
+                       (driver.s/include-schema? schema-inclusion-filters schema-exclusion-filters table-schema)
+                       (i/have-select-privilege? driver conn table-schema table-name)))))
+     (i/syncable-schemas driver conn metadata schema-inclusion-filters schema-exclusion-filters))))
```
This is a bit confusing to me. We filter `(and (driver.s/include-schema? inclusion-filters))` but also, the original seq is `(i/syncable-schemas inclusion-filters)`. It seems we are filtering twice. 

`i/syncable-schemas` for `:sql-jdbc`
```clojure
(defmethod i/syncable-schemas :sql-jdbc
  [driver _ metadata schema-inclusion-patterns schema-exclusion-patterns]
  (eduction (remove (set (i/excluded-schemas driver)))
            (filter (partial driver.s/include-schema? schema-inclusion-patterns schema-exclusion-patterns))
            (all-schemas metadata)))
```
which removes and includes.

For redshift, 
```clojure
(defmethod sql-jdbc.sync/syncable-schemas :redshift
  [driver conn metadata schema-inclusion-patterns schema-exclusion-patterns]
  (reducible-schemas-with-usage-permissions
   conn
   (eduction
    (remove (set (sql-jdbc.sync/excluded-schemas driver)))
    (filter (partial driver.s/include-schema? schema-inclusion-patterns schema-exclusion-patterns))
    (sync.describe-database/all-schemas metadata))))
```

So it seems that `i/syncable-schemas` runs the schemas through redshifts select privileges and then again through `:sql-jdbc`'s `i/have-select-privilege? :sql-jdbc` again, as well as the schema filtering. It seems there's a lot going on here, and lots of it twice.

This code is also hard to follow because of potemkin. Redshift implements `sql-jdbc.sync/syncable-schemas :redshift` but otherwise the default is `i/have-select-privilege? :sql-jdbc` because `metabase.driver.sql-jdbc.sync.interface` contains the actual defmethod whereas `metabase.driver.sql-jdbc.sync` potemkin import-vars this defmethod.

jeff303 replied:
Yeah, there was definitely redundant filtering happening in the `:sql-jdbc` impl.  So I removed that.

Also, the `:redshift` impl actually did need the call to `include-schema?` since it was operating on `all-schemas`.  I changed it up to just delegate to the `:sql-jdbc` impl instead, and only apply the `reducible-schemas-with-usage-permissions` on top of those.

camsaul replied:
> This code is also hard to follow because of Potemkin.

`M-.` still jumps to the actual definition regardless of whether there's a Potemkin alias for it or not

dpsutton commented on src/metabase/driver/util.clj:
```diff
@@ -198,6 +198,13 @@
+  [driver]
+  (first (filter (fn [conn-prop]
+                   (= :schema-filters (keyword (:type conn-prop))))
+           (driver/connection-properties driver))))
```
I'm looking at `driver/connection-properties` and i see implementations only for h2, mysql, and postgres. Presumably redshift goes through the postgres version because of the derivation but I'm surprised there's no default behavior or more implementations.

dpsutton replied:
ah, nevermind i see the docstring about the lazy loading driver.

dpsutton reviewed (approved):
Approving but left some notes. I think the code is a bit confusing. Particularly

```clojure
(eduction
 (comp (mapcat (fn [schema]
                 (db-tables driver metadata schema db-name-or-nil)))
       (filter (fn [{table-schema :schema, table-name :name}]
                 (and
                  (driver.s/include-schema? schema-inclusion-filters schema-exclusion-filters table-schema)
                  (i/have-select-privilege? driver conn table-schema table-name)))))
 (i/syncable-schemas driver conn metadata schema-inclusion-filters schema-exclusion-filters))
```

Where we filter to the schemas we want, and then for each schema, get the tables in that schema, and for each table, filter that the schema is what we want. This makes me think that either a) `i/syncable-schemas` returns schemas that might not match the schema filters, b) `(db-tables ... schema)` might return tables that are not in the schema that was passed in, or c) we're filtering again for no reason.

I think we need a comment explaining why or we need to remove the duplicate filtering.

jeff303 committed:
PR feedback

Add docstrings

Change redshift impl of syncable-schemas to call the sql-jdbc version instead

Remove redundant check of inclusion/exclusion from sql-jdbc impl

jeff303 committed:
Merge branch 'master' into filter-schemas-jdbc-drivers

jeff303 committed:
Fix namespace decls

jeff303 committed:
Merge branch 'master' into filter-schemas-jdbc-drivers

camsaul commented on src/metabase/driver/sql_jdbc/sync/interface.clj:
```diff
@@ -32,8 +35,14 @@
 (defmulti syncable-schemas
   "Return a reducible sequence of string names of schemas that should be synced for the given database. Schemas for
   which the current DB user has no `SELECT` permissions should be filtered out. The default implementation will fetch
-  a sequence of all schema names from the JDBC database metadata and filter out any schemas in `excluded-schemas`."
-  {:added "0.39.0", :arglists '([driver ^java.sql.Connection connection ^java.sql.DatabaseMetaData metadata])}
+  a sequence of all schema names from the JDBC database metadata and filter out any schemas in `excluded-schemas`, along
+  with any that shouldn't be included based on the given inclusion and exclusion patterns (see the
+  `metabase.driver.sync` namespace for full explanation)."
+  {:added "0.39.0", :arglists '([driver
+                                 ^java.sql.Connection connection
+                                 ^java.sql.DatabaseMetaData metadata
+                                 ^String schema-inclusion-patterns
+                                 ^String schema-exclusion-patterns])}
```
What happens to all the drivers that implemented the 3-arity version of this? 💀 

Kind of annoying but if we want to be responsible and developer-friendly we really ought to just make this a new method like we've done in similar situations in the past.

At any rate, changes also need to be mentioned in https://github.com/metabase/metabase/blob/master/docs/developers-guide/driver-changelog.md... since the file actually exists now it's probably best to do it as part of the PR so it can be reviewed together with the other changes

jeff303 replied:
Fixed this, if you can have a look at the update.

camsaul reviewed (changes requested):
- Changing the signature of a driver multimethod will break any 3rd party drivers that implement that method. Since we want to be somewhat developer friendly it's probably better just to make a new method and deprecate the old one
- Now that we have a "what's new for driver authors in ____" page in the actual docs it would be nice to add changes to that doc as part of this PR
- You should invoke "parent methods" of multimethods using `((get-method ...) driver ...)` for reasons mentioned in https://github.com/metabase/metabase/pull/19651#discussion_r792007895

camsaul commented on modules/drivers/redshift/src/metabase/driver/redshift.clj:
```diff
@@ -272,9 +271,7 @@
 
 (defmethod sql-jdbc.sync/syncable-schemas :redshift
-  [driver conn metadata]
+  [_ conn metadata schema-inclusion-patterns schema-exclusion-patterns]
```
```suggestion
  [driver conn metadata schema-inclusion-patterns schema-exclusion-patterns]
```

jeff303 replied:
Fixed in the update.

camsaul commented on modules/drivers/redshift/src/metabase/driver/redshift.clj:
```diff
@@ -272,9 +271,7 @@
-   (eduction
-    (remove (set (sql-jdbc.sync/excluded-schemas driver)))
-    (sync.describe-database/all-schemas metadata))))
+   (sql-jdbc.sync/syncable-schemas :sql-jdbc conn metadata schema-inclusion-patterns schema-exclusion-patterns)))
```
```suggestion
   ((get-method sql-jdbc.sync/syncable-schemas :sql-jdbc) driver conn metadata schema-inclusion-patterns schema-exclusion-patterns)))
```

camsaul replied:
You **definitely** should be invoking "parent methods" this way. Why? Because if `syncable-schemas` in turn calls some **other** multimethod, it's going to use the `driver` parameter you passed it for dispatch. If you pass it `:sql-jdbc`, then it's using the `:sql-jdbc` version of any methods it calls, even if there's a `:redshift`-specific override. So best practice here is to manually get the `:sql-jdbc` version of the method and then pass in the `driver` parameter you got.

Why not pass in `:redshift` directly instead of `driver`? Well, we don't know that someone isn't extending the Redshift driver in some way. This method might actually be getting invoked for some `:redshift-2` driver or something like that. So it's important to pass the actual `driver` we got, not what we think it is. (A good example of this would be the `:postgres` driver. The Postgres driver has to be explicit to pass around `driver` rather than `:postgres`, or else it would skip over Redshift's implementations of things)

jeff303 replied:
Thanks, learned about this in the past but somehow forgot this time.

jeff303 committed:
PR feedback

Use new multimethod instead for filtered-syncable-schemas, and have default impl of syncable-schemas call that

Fix invocation of parent method in :redshift method

jeff303 committed:
Add entries to `driver-changelog.md`

camsaul reviewed (approved):
Awesome. Changes look  good to me
